Lib/interpreters.py

```diff
@@ -0,0 +1,240 @@
+"""Subinterpreters High Level Module."""
+
+import _interpreters
+import logger
+
+__all__ = ['Interpreter', 'SendChannel', 'RecvChannel', 'is_shareable',
+           'create_channel', 'list_all_channels', 'list_all', 'get_current',
+           'create']
+
+
+class Interpreter:
+
+    def __init__(self, id):
+        self.id = id
+
+    def is_running(self):
+        """is_running() -> bool
+
+        Return whether or not the identified interpreter is running.
+        """
+        return _interpreters.is_running(self.id)
+
+    def destroy(self):
+        """destroy()
+
+        Destroy the identified interpreter.
+
+        Attempting to destroy the current
+        interpreter results in a RuntimeError. So does an unrecognized ID
+        """
+        return _interpreters.destroy(self.id)
+
+    def run(self, src_str, /, *, channels=None):
+        """run(src_str, /, *, channels=None)
+
+        Run the given source code in the interpreter.
+        This blocks the current thread until done.
+        """
+        try:
+            _interpreters.run_string(self.id, src_str)
+        except RunFailedError as err:
+            logger.error(err)
+            raise
+
+def wait(self, timeout):
+    #The implementation for wait
+    # will be non trivial to be useful
+    import time
+    time.sleep(timeout)
+
+def associate_interp_to_channel(id, cid):
+    pass
+
+class RecvChannel:
+
+    def __init__(self, id):
+        self.id = id
+        self.interpreters = _interpreters.list_all()
+
+    def recv(self, timeout=2):
+        """ channel_recv() -> obj
+
+        Get the next object from the channel,
+        and wait if none have been sent.
+        Associate the interpreter with the channel.
+        """
+        obj = _interpreters.channel_recv(self.id)
+        if obj == None:
+            wait(timeout)
+            obj = obj = _interpreters.channel_recv(self.id)
+
+        # Pending: See issue 52 on multi-core python project
+        associate_interp_to_channel(interpId, Cid)
+
+        return obj
+
+    def recv_nowait(self, default=None):
+        """recv_nowait(default=None) -> object
+
+        Like recv(), but return the default instead of waiting.
+        """
+        return _interpreters.channel_recv(self.id)
+
+    def send_buffer(self, obj):
+        """ send_buffer(obj)
+
+        Send the object's buffer to the receiving end of the channel
+        and wait. Associate the interpreter with the channel.
+        """
+        pass
+
+    def send_buffer_nowait(self, obj):
+        """ send_buffer_nowait(obj)
+
+        Like send_buffer(), but return False if not received.
+        """
+        pass
+
+    def release(self):
+        """ release()
+
+        No longer associate the current interpreterwith the channel
+        (on the sending end).
+        """
+        return _interpreters.(self.id)
```

Comment:
Should this say `return _interpreters.channel_release(self.id)`? Currently a syntax error.

Reply:
Good catch. Am working on it.